### PR TITLE
Added AsyncEventHandler and extension for safe invocation

### PR DESCRIPTION
**Pull request by Khomutov**

### CODE REVIEW

File: src/Catel.Core/Catel.Core.Shared/EventHandlers/Extensions/AsyncEventHandlerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="AsyncEventHandlerExtensions.cs" company="Catel development team">
+//   Copyright (c) 2008 - 2015 Catel development team. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Catel.EventHandlers.Extensions
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/EventHandlers/Extensions/AsyncEventHandlerExtensions.cs]: Fix namespace please

File: src/Catel.Core/Catel.Core.Shared/EventHandlers/AsyncEventHandler.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="AsyncEventHandler.cs" company="Catel development team">
+//   Copyright (c) 2008 - 2015 Catel development team. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Catel.EventHandlers
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/EventHandlers/AsyncEventHandler.cs]: Use the Catel namespace only, remove EventHandlers

File: src/Catel.Core/Catel.Core.Shared/EventHandlers/Extensions/AsyncEventHandlerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        /// <summary>
+        /// Invokes the specified <paramref name="handler"/>
+        /// </summary>
+        /// <param name="handler">The handler.</param>
+        /// <param name="sender">The sender.</param>
+        /// <param name="e">The event args.</param>
+        /// <typeparam name="TEventArgs">The type of the <see cref="EventArgs"/> class.</typeparam>
+        /// <returns></returns>
+        public static async Task<bool> SafeInvoke<TEventArgs>(this AsyncEventHandler<TEventArgs> handler, object sender, TEventArgs e) where TEventArgs : EventArgs
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/EventHandlers/Extensions/AsyncEventHandlerExtensions.cs]: where TEventArgs : EventArgs on a new line please

File: src/Catel.Core/Catel.Core.Shared/EventHandlers/Extensions/AsyncEventHandlerExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            if (handler == null)
+            {
+                return false;
+            }
+
+            var eventListeners = handler.GetInvocationList().Cast<AsyncEventHandler<TEventArgs>>();
+            foreach (var eventListener in eventListeners)
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/EventHandlers/Extensions/AsyncEventHandlerExtensions.cs]: Don't use foreach. People might change subscriptions inside a handler => boom. See https://github.com/Catel/Catel/blob/c29755ba1574b2b47020b561149e021175d14b5c/src/Catel.Core/Catel.Core.Shared/EventHandlerExtensions.cs for a decent implementation.

**Comment by Khomutov** [comment]
fixed

**Pull request merged**